iceoryx_utils/include/iceoryx_utils/internal/concurrent/lockfree_queue/resizeable_lockfree_queue.inl

```diff
@@ -0,0 +1,233 @@
+// Copyright (c) 2020 by Apex.AI Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "iceoryx_utils/concurrent/resizeable_lockfree_queue.hpp"
+
+namespace iox
+{
+namespace concurrent
+{
+template <typename ElementType, uint64_t MaxCapacity>
+ResizeableLockFreeQueue<ElementType, MaxCapacity>::ResizeableLockFreeQueue(uint64_t initialCapacity) noexcept
+{
+    setCapacity(initialCapacity);
+}
+
+template <typename ElementType, uint64_t MaxCapacity>
+constexpr uint64_t ResizeableLockFreeQueue<ElementType, MaxCapacity>::maxCapacity() noexcept
+{
+    return MAX_CAPACITY;
+}
+
+template <typename ElementType, uint64_t MaxCapacity>
+uint64_t ResizeableLockFreeQueue<ElementType, MaxCapacity>::capacity() const noexcept
+{
+    return m_capacity.load(std::memory_order_relaxed);
+}
+
+
+template <typename ElementType, uint64_t MaxCapacity>
+template <typename ContainerType, typename>
+bool ResizeableLockFreeQueue<ElementType, MaxCapacity>::setCapacity(uint64_t newCapacity,
+                                                                    ContainerType& removedElements) noexcept
+{
+    //@todo: preferably the signature should be void(ElementType&) but this leads to a compile error
+    //       reason for change is to reflect the fact that the elements may be changed
+    //       by the removeHandler (main use case is move)
+    auto removeHandler = [&](const ElementType& value) { removedElements.emplace_back(std::move(value)); };
+    return setCapacity(newCapacity, removeHandler);
+}
+
+template <typename ElementType, uint64_t MaxCapacity>
+bool ResizeableLockFreeQueue<ElementType, MaxCapacity>::setCapacity(uint64_t newCapacity) noexcept
+{
+    auto removeHandler = [](const ElementType&) {};
+    return setCapacity(newCapacity, removeHandler);
+}
+
+template <typename ElementType, uint64_t MaxCapacity>
+template <typename Function, typename>
+bool ResizeableLockFreeQueue<ElementType, MaxCapacity>::setCapacity(uint64_t newCapacity,
+                                                                    Function&& removeHandler) noexcept
+{
+    if (newCapacity > MAX_CAPACITY)
+    {
+        return false;
+    }
+
+    if (m_resizeInProgress.test_and_set(std::memory_order_acquire))
```

Comment:
Please do not hide atomic call in methods in such complex lock-free code. Could you please remove `test_and_set` and put the code directly in here so that I have a chance in github to see the whole algorithm.

Reply:
to remove `test_and_set` you would have to talk to the C++ committee ;)

Reply:
hehehe whoopsie ... forgot that this is a `atomic_flag`.

Reply:
This guards against multiple setCapacity operations interfering with the vector but does not block progress. There is a secret way around this, but I think I cannot use this. :-(